opentelemetry-sdk/src/logs/log_processor.rs

```diff
@@ -994,4 +998,223 @@ mod tests {
+        fn len(&self) -> usize {
+            self.event_count.load(Ordering::Acquire) as usize
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl LogExporter for LogExporterThatRequiresTokio {
+        async fn export(&mut self, batch: LogBatch<'_>) -> LogResult<()> {
+            // Simulate minimal dependency on tokio by sleeping asynchronously for a short duration
+            tokio::time::sleep(Duration::from_millis(50)).await;
+
+            for _ in batch.iter() {
+                self.event_count.fetch_add(1, Ordering::Acquire);
+            }
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn test_simple_processor_async_exporter_without_runtime() {
+        // Use `catch_unwind` to catch the panic caused by missing Tokio runtime
+        let result = std::panic::catch_unwind(|| {
+            let exporter = LogExporterThatRequiresTokio::new();
+            let processor = SimpleLogProcessor::new(Box::new(exporter.clone()));
+
+            let mut record: LogRecord = Default::default();
+            let instrumentation: InstrumentationLibrary = Default::default();
+
+            // This will panic because an tokio async operation within exporter without a runtime.
+            processor.emit(&mut record, &instrumentation);
+        });
+
+        // Verify that the panic occurred and check the panic message for the absence of a Tokio runtime
+        assert!(
+            result.is_err(),
+            "The test should fail due to missing Tokio runtime, but it did not."
+        );
+        let panic_payload = result.unwrap_err();
+        let panic_message = panic_payload
+            .downcast_ref::<String>()
+            .map(|s| s.as_str())
+            .or_else(|| panic_payload.downcast_ref::<&str>().copied())
+            .unwrap_or("No panic message");
+
+        assert!(
+            panic_message.contains("no reactor running")
+                || panic_message.contains("must be called from the context of a Tokio 1.x runtime"),
+            "Expected panic message about missing Tokio runtime, but got: {}",
+            panic_message
+        );
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+    async fn test_simple_processor_async_exporter_with_runtime() {
+        let exporter = LogExporterThatRequiresTokio::new();
+        let processor = SimpleLogProcessor::new(Box::new(exporter.clone()));
+
+        let mut record: LogRecord = Default::default();
+        let instrumentation: InstrumentationLibrary = Default::default();
+
+        processor.emit(&mut record, &instrumentation);
+
+        assert_eq!(exporter.len(), 1);
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+    #[ignore]
+    // All worker threads except one are blocked, waiting for the export operation to complete.
+    // The exporter, which isn't blocked, requires the runtime to proceed, but no free worker threads are available, resulting in a deadlock.
```

Comment:
Let's a provide a more specific reasoning here? How about something like this?

We spawn enough number of tasks to have each worker thread execute export. While executing export, each of these worker threads park themselves when they hit tokio `sleep` (since `sleep` is called within `futures_executor::block_on`). We need a worker thread to react to the completion of sleep and unpark these threads but there is none available. As a result, the worker threads never get unparked and the spawned tokio tasks don't ever complete.

Reply:
@utpilla actually only one worker thread is blocked on the tokio `sleep`, rest are blocked on the `mutex` for processor. I have added the comment to make it clear as below:
```
    // This test demonstrates a potential deadlock scenario in a multi-threaded Tokio runtime.
    // It spawns Tokio tasks equal to the number of runtime worker threads (4) to emit log events.
    // Each task attempts to acquire a mutex on the SimpleLogProcessor. Only one task obtains the lock,
    // while the others are blocked, waiting for its release.
    //
    // The task holding the lock invokes the LogExporterThatRequiresTokio, which performs an
    // asynchronous operation (e.g., network I/O simulated by `tokio::sleep`). This operation
    // requires yielding control back to the Tokio runtime to make progress.
    //
    // However, all worker threads are occupied:
    // - One thread is executing the async exporter operation
    // - Three threads are blocked waiting for the mutex
    //
    // This leads to a deadlock as there are no available threads to drive the async operation
    // to completion, preventing the mutex from being released. Consequently, neither the blocked
    // tasks nor the exporter can proceed.
```

Reply:
I see. Just a general note, I feel that the `Mutex` used by `SimpleLogProcessor` is hiding away the main issue here which is `futures_executor::block_on`. For this particular test, yes since we use a `SimpleLogProcessor` other worker threads are waiting on the `Mutex`. That might lead us into thinking that `Mutex` is the real issue here but it's not.

Even if we were to remove the `Mutex`, and use something like a Reentrant/no-lock processor we would still hit this issue as long as we have `futures_exectuor::block_on`. Try this code which takes no mutex and spawns a task on each worker thread to wait for some time. This would still hang due to the reasoning stated in this [comment](https://github.com/open-telemetry/opentelemetry-rust/pull/2218#discussion_r1807055338).

```rust
#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn test_simple_processor_async_exporter_with_all_runtime_worker_threads_blocked() {
        let concurrent_emit = 4; // number of worker threads

        let mut handles = vec![];
        for _ in 0..concurrent_emit {
            let handle = tokio::spawn(async move {
                futures_executor::block_on(async { tokio::time::sleep(Duration::from_millis(50)).await });
            });
            handles.push(handle);
        }

        for handle in handles {
            handle.await.unwrap();
        }
}

Reply:
+1 to above.
I left https://github.com/open-telemetry/opentelemetry-rust/pull/2218#discussion_r1808032002 to mean same thing, but above is a very good explanation!

Reply:
@utpilla In case of SimpleLogProcessor, we will never have concurrent execution of `block_on` because of the mutex here - https://github.com/open-telemetry/opentelemetry-rust/blob/84c151c0b6f755b71c466f98d54a6336ebe159aa/opentelemetry-sdk/src/logs/log_processor.rs#L110

So actually it is the combination of mutex and block_on in the processor code that is causing deadlock :)

However, I have removed the mutex from the unit-test, as there is already mutex as above, as also you indicated here - https://github.com/open-telemetry/opentelemetry-rust/pull/2218#discussion_r1808080261. Hopefully this will make it more clear now.

@cijothomas  - In this test now that we don't lock processor, we invoke SimpleLogProcessor::emit concurrently from multiple tokio worker threads, which would be the similar scenario when the user invokes multiple tracing macros concurrently.